docs/source/how-to/custom-formats.md

```diff
@@ -13,18 +13,156 @@ assumptions about the array.)
 
 For this guide, we'll take the example of
 [XDI](https://github.com/XraySpectroscopy/XAS-Data-Interchange/blob/master/specification/spec.md#example-xdi-file),
-which is a formalized text-based format for X-ray Spectroscopy data. As you can
-see from the linked example, the file's contents comprise a single table
+which is a formalized text-based format for X-ray Spectroscopy data.
+An example XDI file is reproduced below.
+```
+# XDI/1.0 GSE/1.0
+# Column.1: energy eV
+# Column.2: i0
+# Column.3: itrans
+# Column.4: mutrans
+# Element.edge: K
+# Element.symbol: Cu
+# Scan.edge_energy: 8980.0
+# Mono.name: Si 111
+# Mono.d_spacing: 3.13553
+# Beamline.name: 13ID
+# Beamline.collimation: none
+# Beamline.focusing: yes
+# Beamline.harmonic_rejection: rhodium-coated mirror
+# Facility.name: APS
+# Facility.energy: 7.00 GeV
+# Facility.xray_source: APS Undulator A
+# Scan.start_time: 2001-06-26T22:27:31
+# Detector.I0: 10cm  N2
+# Detector.I1: 10cm  N2
+# Sample.name: Cu
+# Sample.prep: Cu metal foil
+# GSE.EXTRA:  config 1
+# ///
+# Cu foil Room Temperature
+# measured at beamline 13-ID
+#----
+# energy i0 itrans mutrans
+  8779.0  149013.7  550643.089065  -1.3070486
+  8789.0  144864.7  531876.119084  -1.3006104
+  8799.0  132978.7  489591.10592  -1.3033816
+  8809.0  125444.7  463051.104096  -1.3059724
+  8819.0  121324.7  449969.103983  -1.3107085
+  8829.0  119447.7  444386.117562  -1.3138152
+  8839.0  119100.7  440176.091039  -1.3072055
+  8849.0  117707.7  440448.106567  -1.3195882
+  8859.0  117754.7  442302.10637  -1.3233895
+  8869.0  117428.7  441944.116528  -1.3253521
+  8879.0  117383.7  442810.120466  -1.327693
+  8889.0  117185.7  443658.11566  -1.3312944
+```
+As you can see , the file's contents comprise a single table
 and a header of key--value pairs. Therefore, it makes sense to export `dataframe`
 structures and their associated metadata into this format.
 
-Take the following simple server configuration:
+Consider a folder `data` containing a number of xdi files that we wish to serve using tiled.
+An example folder containing a single `example.xdi` file can be generated by running `python -m tiled.examples.xdi`.
+```{note}
+All code below is in `tiled/examples/xdi.py`.
+The code is reproduced for reference but running the examples only requires modifying `config.yaml`.
+```
+
+# Define an Importer
+
+The first step in exposing this data via tiled is to write a function which can read xdi formatted files.
+Note this function is provided as an example and does not do the validation that would be expected from a full parser.
+
+```py
+# tiled/examples/xdi.py
+
+def read_xdi(path):
+    with open(path, "r") as f:
+        metadata = {}
+        fields = collections.defaultdict(dict)
+
+        line = f.readline()
+        m = re.match("# XDI/(\S*)\s*(\S*)?", line)
+        if not m:
+            raise AssertionError(f"not an XDI file, no XDI versioning information in first line\n{line}")
```

Comment:
This should be `ValueError`. CPython says assertions are for "This state should be impossible; the program has a bug," not for input validation.